README.md
# Carcassonne Score keeping

Score keeping software for [Carcassonne](https://boardgamegeek.com/boardgame/822/carcassonne) to facilitate statistical analysis of games.
Information (players, games, scores, expansions) are stored in a sqlite3 database.
The location of this database defaults to `CarcassonneScore.db`, but can be specified in the `CarcassonneScore.conf` configuration file.

## Requirements

Detailed information in [`requirements.txt`](requirements.txt)

* python 3
* numpy
* matplotlib (only needed for for analysis scripts)

## Usage

Before your first game you will need to initialize the sqlite database:

```
$ python manage_database.py --init
```

### Score keeping

After a database has been initialized, you can score a game with:

```
$ python CarcassonneScore.py
```

This launches the interactive shell which sets up a game (gets the names and order of players, list of active expansions, and other information).
Within this shell various commands are available to record scores, advance to a next player's turn, and end the game.
Press `?` within the shell for a list of commands (note that the list of available commands may change depending on the expansions being played and whether the game is in active tile-placing mode or post-game scoring mode).

### Database Maintenance

To update the database (e.g., to add new players or toggle availability of expansions), use the `manage_database.py` command. For example, to add a new player:

```
$ python manage_database.py -n NEW PLAYER
```

Use `python manage_database.py -h` to see the full list of options.

### Analysis

Some jupyter notebooks with sample analysis and a sample sqlite database containing one game are available in the `analysis/` directory.

### Miscellaneous Utilities

The `utilities/` directory contains scripts which may be useful for correcting issues, but these should not be needed for normal operations.


CHANGELOG.md
# Changelog

## 0.5 Series

### 0.5.0 (in progress)

#### Enhancements

* Add ability to (de)activate players.
  **NOTE**: If you have been using an earlier version, this requires running `manage_database.py --migrate` to update the database.

## 0.4 Series

### 0.4.4 (07 December 2020)

#### Enhancements

* Note when a score has not been recorded.

### 0.4.3 (22 May 2019)

#### Enhancements

* Print token type in scoring confirmation message. Update formatting.
* Add "Hills and & Sheep" Expansion.
  To update your database, run this SQLITE statement:
  `update expansions set active=1, tokens="Shepherd", scoretypes="Flock", Ntiles=18, tiletypes="Hill,Vineyard" where expansionID=9;`
* At the end of play and the end of final scoring, print score sorted from highest to lowest score.

#### Bug Fixes

* Update single-game and score distribution demo scripts to generalize color cycler and hatching for larger numbers of players.

### 0.4.2 (21 November 2018)

#### Enhancements

* Add turn length vs turn number plot to `analysis/SampleAnalysis-SingleGame.ipynb`.
* Reorder game initilization so that most of the info can be entered before the player order is determined.
* Announce gameID and location at the start.
* Status printout now shows time elapsed since end of previous turn.
* In the status message, showing the number of tiles played/remaining and the time elapsed since the end of the previous turn is now controlled via options in the configuration file.
* Add `requirements.txt` file.
* Print the player's name when confirming a score entry.

#### Bug fixes

* Fixed tile counting bug when an Abbey is played.
* Fixed missing Abbey score type when Expansion 5 is used.
  This requres running the following command in sqlite3: `update expansions set scoretypes="Abbey" where expansionID=5;`
* Include played Abbey tiles in the total number of tiles played.
* Ensure negative scores cannot be entered.
* Print all players who are scoring when confirming.

### 0.4.1 (18 November 2018)

#### Enhancements

* Add `analysis/README.md` file to briefly explain what the analysis scripts do.
* Add some additional explanatory text to the main `README.md` file.
* Add Expansion 3 (The Princes and the Dragon) to SQL database creation.
* Reorder score recording questions to better match how players seem to naturally report scores.
* Add a command to note that a player played an Abbey tile (if "Abbey & Mayor" expansion is in play). This increments the turn as normal but does not increment the number of tiles played (so that the tiles remaining count remains accurate).

#### Bug fixes

* Fixed bug where `ScoreProbByType.ipynb` sample script failed when showing PDFs for trade token score distributions.
* Use `ORDER BY` in some analysis queries to ensure turn/score information is returned in the proper order.
* Minor plotting code updates in analysis to fix depredations in matplotlib.
* Fix off-by-one error in number of tiles remaining which occurs when The River expansion is in use.

### 0.4.0 (11 November 2018)

#### Enhancements

* add a configuration file specifying the database location
* add a utilities directory for scripts which should not be needed for normal operation but may be useful for working around issues.
* add scoring capability for Trade tokens (Traders & Builders Expansion). Token allocation is not automatically tracked (this can be tracked manually using the scoring comments). Trade token scores are entered during the post-game scoring period.

## 0.3 Series

### 0.3.4 (28 February 2018)

#### Enhancements

* analysis script to look at a player's scoring trends with time

### 0.3.3 (21 February 2018)

#### Enhancements

* Example analysis to show probability distributions of score by score type.

#### Bug fixes

* properly count the starting tile
* fix labeling issue in analysis script
* check validity of player IDs when a score is entered

### 0.3.2 (07 February 2018)

#### Bug fixes

* syntax error in `manage_database.py`

### 0.3.1 (07 February 2018)

#### Bug fixes

* Correct version number in source code

### 0.3.0 (07 February 2018)

#### Enhancements

* new `manage_database.py` script to create and maintain the database
* add database versioning

## 0.2 Series

### 0.2.0 (04 February 2018)

#### Bug fixes

* check inputs

#### Enhancements

* Require a minimum of 2 players
* Implement a check on scores in case bad values are entered
* Sample game data and an analysis notebook in the `analysis/` directory

## 0.1 Series

### 0.1.1 (03 February 2018)

#### Bug fixes

* Add missing score type

### 0.1.0 (03 February 2018)

Initial release.



create_database.py
#!/usr/bin/env python
"""
Database initialization for Carcassonne score keeping system.

Copyright 2018 George C. Privon

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

import sqlite3

conn = sqlite3.connect('CarcassonneScore.db')
c = conn.cursor()

# player table
c.execute('''CREATE TABLE players (playerID INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
                                   name TEXT NOT NULL DEFAULT "")''')
# player names
c.execute('''INSERT INTO players values (0,
                                         "John Smith")''')
c.execute('''INSERT INTO players values (1,
                                         "Jane Doe")''')

# games table
# gameID - unique ID
# location - free text
# starttime - timecode
# endtime - timecode
# expansions - comma-separated list of expansions used (by expansionID)
c.execute('''CREATE TABLE games (gameID INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
                                 location TEXT NOT NULL DEFAULT "",
                                 starttime TEXT NOT NULL DEFAULT "",
                                 endtime TEXT NOT NULL DEFAULT "",
                                 expansions TEXT NOT NULL DEFAULT "")''')


# turns table
# gameID - corresponds to games table
# turnNum - incremented turn number (corresponds to a tile placement)
# time - time at which the placement was made
# builder - 0 if tile was placed as part of normal turn, 1 if placed
#           as a result of a builder token
# playerID - player who made the turn
c.execute('''CREATE TABLE turns (gameID INTEGER NOT NULL,
                                 turnNum INTEGER NOT NULL,
                                 time TEXT NOT NULL,
                                 builder INTEGER NOT NULL,
                                 playerID INTEGER NOT NULL)''')

# individual scores
# gameID - unique per game
# playerID - corresponds to entry from players table
# turnNum - turn number (corresponds to number in turns table)
# scoreID - ID number for the score (unique to game?)
# ingame - 1 if scored during regular play, 0 if scored after tiles are gone
# points - number of points awarded
# scoretype - road, city, meadow, etc.
# sharedscore - was this score shared with another player?
# token - token(s) used for score (e.g., meeple, wagon, pig)
# extras - any other items (e.g., trade goods)
# comments - text annotation
c.execute('''CREATE TABLE scores (gameID INTEGER NOT NULL,
                                  playerID INTEGER NOT NULL,
                                  turnNum INTEGER NOT NULL,
                                  scoreID INTEGER NOT NULL,
                                  ingame INTEGER NOT NULL,
                                  points INTEGER NOT NULL,
                                  scoretype TEXT NOT NULL,
                                  sharedscore INTEGER NOT NULL,
                                  token TEXT NOT NULL,
                                  extras NOT NULL,
                                  comments TEXT)''')

# list of expansions
# mini - 1 if a mini expanion, otherwise 0
# active - 1 if it can be played, otherwise 0
# tokens - additional tokens provided beyond the base game
# scoretypes - additional classes of scoring enabled
# Ntiles - number of tiles added
# tiletypes - special scorable tiles added
c.execute('''CREATE TABLE expansions (expansionID INTEGER PRIMARY KEY,
                                      name TEXT NOT NULL,
                                      mini INTEGER,
                                      active INTEGER,
                                      tokens TEXT,
                                      scoretypes TEXT,
                                      Ntiles INTEGER,
                                      tiletypes TEXT)''')
# large expansions
# taken from http://carcassonne.wikia.com/wiki/Official_expansions
# Only 1, 2, and 5 are currently populated
c.execute('''INSERT INTO expansions values (1,
                                            "Inns & Cathedrals",
                                            0,
                                            1,
                                            "BigMeeple",
                                            "",
                                            18,
                                            "Cathedral,InnonLake")''')
c.execute('''INSERT INTO expansions values (2,
                                            "Traders & Builders",
                                            0,
                                            1,
                                            "Pig,Builder",
                                            "",
                                            24,
                                            "TradeGoods")''')
c.execute('''INSERT INTO expansions values (3,
                                            "Princess & Dragon",
                                            0,
                                            0,
                                            "",
                                            "",
                                            0,
                                            "")''')
c.execute('''INSERT INTO expansions values (4,
                                            "The Tower",
                                            0,
                                            0,
                                            "",
                                            "",
                                            0,
                                            "")''')
c.execute('''INSERT INTO expansions values (5,
                                            "Abbey & Mayor",
                                            0,
                                            1,
                                            "Mayor,Wagon,Barn",
                                            "",
                                            12,
                                            "Abbey")''')
c.execute('''INSERT INTO expansions values (6,
                                            "Count, King & Robber",
                                            0,
                                            0,
                                            "",
                                            "",
                                            0,
                                            "")''')
c.execute('''INSERT INTO expansions values (7,
                                            "The Catapult",
                                            0,
                                            0,
                                            "",
                                            "",
                                            0,
                                            "")''')
c.execute('''INSERT INTO expansions values (8,
                                            "Bridges, Castles & Bazaars",
                                            0,
                                            0,
                                            "",
                                            "",
                                            0,
                                            "")''')
c.execute('''INSERT INTO expansions values (9,
                                            "Hills & Sheep",
                                            0,
                                            0,
                                            "",
                                            "",
                                            0,
                                            "")''')
c.execute('''INSERT INTO expansions values (10,
                                            "Under the Big Top",
                                            0,
                                            0,
                                            "",
                                            "",
                                            0,
                                            "")''')
#mini expansions
c.execute('''INSERT INTO expansions values(101,
                                           "The River",
                                           1,
                                           1,
                                           "",
                                           "",
                                           12,
                                           "")''')
c.execute('''INSERT INTO expansions values(102,
                                           "The Abbott",
                                           1,
                                           1,
                                           "Abbott",
                                           "Garden",
                                           0,
                                           "")''')

conn.commit()
conn.close()


CarcassonneScore.py
#!/usr/bin/env python
"""
Carcassonne score keeping system.

Copyright 2018-2019 George C. Privon

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

import argparse
import sys
import cgame

__version__ = "0.4.4"

def getargs():
    """
    Get command line arguments.
    """

    parser = argparse.ArgumentParser(description="Carcassonne score keeping \
system.")
    parser.add_argument('-c', '--config', default='CarcassonneScore.conf',
                        help='Location of the configuration file.')
    parser.add_argument('--random-exp', default=False,
                         help="Select a random set of active expansions and \
mini-expansions.")

    return parser.parse_args()


def main():
    """
    Main routine
    """

    args = getargs()

    mygame = cgame.cgame(args=args)

    mygame.runGame()

    sys.stdout.write("Thanks for playing!\n\n")


if __name__ == "__main__":
    main()


manage_database.py
#!/usr/bin/env python
"""
Database management for Carcassonne score keeping system.

Copyright 2018 George C. Privon

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

import sys
import os
import argparse
import configparser
import re
import sqlite3


def loadConfig(cfile):
    """
    Load configuration file
    """

    if not os.path.isfile(cfile):
        sys.stderr.write("Error: could not find configuration file '" + cfile + "'\n")
        if os.path.isfile('CarcassonneScore.conf.example'):
            sys.stderr.write("Copying 'CarcassonneScore.conf.example' to 'CarcassonneScore.conf'.\n")
            import shutil
            shutil.copyfile('CarcassonneScore.conf.example',
                            'CarcassonneScore.conf')
        else:
            sys.stderr.write("Error: cannot find default configuration file 'CarcassonneScore.conf.example' to copy.\n")

    config = configparser.RawConfigParser()
    config.read(cfile)

    return config


def parseArgs():
    """
    Command line arguments
    """

    parser = argparse.ArgumentParser(description="Update the Carcassonne \
scoring database.")
    parser.add_argument('-c', '--config', default='CarcassonneScore.conf',
                        help='Location of the configuration file.')
    cmds = parser.add_mutually_exclusive_group(required=True)
    cmds.add_argument('--init', default=False, action='store_true',
                      help='Create a fresh database.')
    cmds.add_argument('--migrate', default=False, action='store_true',
                      help='Update database to a new version.')
    cmds.add_argument('-n', '--newplayer', type=str, default=None,
                      nargs='+',
                      help='Add a new player.')
    cmds.add_argument('-pa', '--playeractivate', action='store_true',
                      default=False,
                      help='activate a player.')
    cmds.add_argument('-pd', '--playerdeactivate', action='store_true',
                      default=False,
                      help='Deactivate a player.')
    cmds.add_argument('-e', '--enableexpansion', action='store_true',
                      default=False,
                      help='Enable an expansion.')
    cmds.add_argument('-d', '--disableexpansion', action='store_true',
                      default=False,
                      help='Disable an expansion.')

    return parser.parse_args()


def initializeDB(c, DBVER):
    """
    Initialize Database
    """

    # player table
    c.execute('''CREATE TABLE players (playerID INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
                                       name TEXT NOT NULL DEFAULT "")''')
    # player names
    c.execute('''INSERT INTO players values (0,
                                             "John Smith")''')
    c.execute('''INSERT INTO players values (1,
                                             "Jane Doe")''')

    # games table
    # gameID - unique ID
    # location - free text
    # starttime - timecode
    # endtime - timecode
    # expansions - comma-separated list of expansions used (by expansionID)
    c.execute('''CREATE TABLE games (gameID INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
                                     location TEXT NOT NULL DEFAULT "",
                                     starttime TEXT NOT NULL DEFAULT "",
                                     endtime TEXT NOT NULL DEFAULT "",
                                     expansions TEXT NOT NULL DEFAULT "")''')


    # turns table
    # gameID - corresponds to games table
    # turnNum - incremented turn number (corresponds to a tile placement)
    # time - time at which the placement was made
    # builder - 0 if tile was placed as part of normal turn, 1 if placed
    #           as a result of a builder token
    # playerID - player who made the turn
    c.execute('''CREATE TABLE turns (gameID INTEGER NOT NULL,
                                     turnNum INTEGER NOT NULL,
                                     time TEXT NOT NULL,
                                     builder INTEGER NOT NULL,
                                     playerID INTEGER NOT NULL)''')

    # individual scores
    # gameID - unique per game
    # playerID - corresponds to entry from players table
    # turnNum - turn number (corresponds to number in turns table)
    # scoreID - ID number for the score (unique to game?)
    # ingame - 1 if scored during regular play, 0 if scored after tiles are gone
    # points - number of points awarded
    # scoretype - road, city, meadow, etc.
    # sharedscore - was this score shared with another player?
    # token - token(s) used for score (e.g., meeple, wagon, pig)
    # extras - any other items (e.g., trade goods)
    # comments - text annotation
    c.execute('''CREATE TABLE scores (gameID INTEGER NOT NULL,
                                      playerID INTEGER NOT NULL,
                                      turnNum INTEGER NOT NULL,
                                      scoreID INTEGER NOT NULL,
                                      ingame INTEGER NOT NULL,
                                      points INTEGER NOT NULL,
                                      scoretype TEXT NOT NULL,
                                      sharedscore INTEGER NOT NULL,
                                      token TEXT NOT NULL,
                                      extras NOT NULL,
                                      comments TEXT)''')

    # list of expansions
    # mini - 1 if a mini expanion, otherwise 0
    # active - 1 if it can be played, otherwise 0
    # tokens - additional tokens provided beyond the base game
    # scoretypes - additional classes of scoring enabled
    # Ntiles - number of tiles added
    # tiletypes - special scorable tiles added
    c.execute('''CREATE TABLE expansions (expansionID INTEGER PRIMARY KEY,
                                          name TEXT NOT NULL,
                                          mini INTEGER,
                                          active INTEGER,
                                          tokens TEXT,
                                          scoretypes TEXT,
                                          Ntiles INTEGER,
                                          tiletypes TEXT)''')
    # large expansions
    # taken from http://carcassonne.wikia.com/wiki/Official_expansions
    # Only 1, 2, and 5 are currently populated
    c.execute('''INSERT INTO expansions values (1,
                                                "Inns & Cathedrals",
                                                0,
                                                1,
                                                "BigMeeple",
                                                "",
                                                18,
                                                "Cathedral,InnonLake")''')
    c.execute('''INSERT INTO expansions values (2,
                                                "Traders & Builders",
                                                0,
                                                1,
                                                "Pig,Builder",
                                                "",
                                                24,
                                                "TradeGoods")''')
    c.execute('''INSERT INTO expansions values (3,
                                                "Princess & Dragon",
                                                0,
                                                0,
                                                "Dragon,Fairy",
                                                "",
                                                30,
                                                "")''')
    c.execute('''INSERT INTO expansions values (4,
                                                "The Tower",
                                                0,
                                                0,
                                                "",
                                                "",
                                                0,
                                                "")''')
    c.execute('''INSERT INTO expansions values (5,
                                                "Abbey & Mayor",
                                                0,
                                                1,
                                                "Mayor,Wagon,Barn",
                                                "Abbey",
                                                12,
                                                "Abbey")''')
    c.execute('''INSERT INTO expansions values (6,
                                                "Count, King & Robber",
                                                0,
                                                0,
                                                "",
                                                "",
                                                0,
                                                "")''')
    c.execute('''INSERT INTO expansions values (7,
                                                "The Catapult",
                                                0,
                                                0,
                                                "",
                                                "",
                                                0,
                                                "")''')
    c.execute('''INSERT INTO expansions values (8,
                                                "Bridges, Castles & Bazaars",
                                                0,
                                                0,
                                                "",
                                                "",
                                                0,
                                                "")''')
    c.execute('''INSERT INTO expansions values (9,
                                                "Hills & Sheep",
                                                0,
                                                0,
                                                "Shepherd",
                                                "Flock",
                                                18,
                                                "Hill,Vineyard")''')
    c.execute('''INSERT INTO expansions values (10,
                                                "Under the Big Top",
                                                0,
                                                0,
                                                "",
                                                "",
                                                0,
                                                "")''')
    #mini expansions
    c.execute('''INSERT INTO expansions values(101,
                                               "The River",
                                               1,
                                               1,
                                               "",
                                               "",
                                               12,
                                               "")''')
    c.execute('''INSERT INTO expansions values(102,
                                               "The Abbott",
                                               1,
                                               1,
                                               "Abbott",
                                               "Garden",
                                               0,
                                               "")''')

    c.execute('PRAGMA user_version={0:1.0f}'.format(DBVER))


def getPlayers(cur, active=True):
    """
    Get a list of (in)active expansions
    """

    command = 'SELECT playerID,name FROM players WHERE active='
    if active:
        command = command + '1'
    else:
        command = command + '0'

    return cur.execute(command).fetchall()


def togglePlayer(cur, ID, activate=True):
    """
    Change the active status for a player
    """
    command = 'UPDATE players SET active='
    if activate:
        command = command + '1'
    elif not activate:
        command = command + '0'
    else:
        sys.stderr.write("What have you done?\n")
        sys.exit(-1)

    command = command + ' where playerID={0:1.0f}'.format(ID)

    cur.execute(command)


def getExpans(cur, active=True):
    """
    Get a list of (in)active expansions
    """

    command = 'SELECT expansionID,name FROM expansions WHERE active='
    if active:
        command = command + '1'
    else:
        command = command + '0'

    return cur.execute(command).fetchall()


def toggleExpan(cur, ID, activate=True):
    """
    Change the active status for an expansion
    """
    command = 'UPDATE expansions SET active='
    if activate:
        command = command + '1'
    elif not activate:
        command = command + '0'
    else:
        sys.stderr.write("What have you done?\n")
        sys.exit(-1)

    command = command + ' where expansionID={0:1.0f}'.format(ID)

    cur.execute(command)


def migratedb(cur, newDBVER):
    """
    Automatically migrate a Carcassonne sqlite database to a newer schema
    """

    # get USER_VERSION from the Sqlite3 database
    curDBVER = cur.execute("PRAGMA user_version").fetchall()[0][0]

    # Check if it is the same as the new version
    # - If it is, print a message that no migration is needed, then exit
    if curDBVER == newDBVER:
        sys.stderr.write("Database schema is up to date, no migration needed.\n")
        sys.exit()

    # If migration is needed, step through each upgrade step and execute it
    if curDBVER == 0:
        # make changes from DBVER 0 to DBVER 1
        cur.execute("ALTER TABLE players ADD COLUMN active INTEGER NOT NULL DEFAULT 1;")
        cur.execute("PRAGMA user_version={0:d}".format(1))
        curDBVER = 1


def main():
    """
    main routine
    """

    args = parseArgs()

    copts = loadConfig(args.config)

    DBNAME = 'CarcassonneScore.db'
    DBVER = 1

    if args.init and os.path.isfile(DBNAME):
        sys.stderr.write("Error: '" + DBNAME + "' already exists. Exiting.\n")
        sys.exit()

    conn = sqlite3.connect(DBNAME)
    cur = conn.cursor()

    VALID = False

    if args.init:
        # create initial database
        initializeDB(cur, 0)
        # migrate it to the current version
        migratedb(cur, DBVER)
    if args.migrate:
        migratedb(cur, DBVER)
    elif args.newplayer:
        pname = ' '.join(args.newplayer)
        sys.stdout.write("Adding new player: " + pname)
        sys.stdout.write(" to the database.\n")
        while not VALID:
            ans = input("Is this correct (y/n)? ")
            if re.match('y', ans, re.IGNORECASE):
                cur.execute('INSERT INTO players (name) VALUES ("' + \
                            pname + '")')
                sys.stdout.write(pname + ' added to the database.\n')
                VALID = True
            elif re.match('n', ans, re.IGNORECASE):
                sys.stdout.write('Canceling.\n')
                VALID = True
    elif args.playeractivate:
        players = getPlayers(cur, active=False)
        kwds = ("enable", "inactive")
        activate = True
    elif args.playerdeactivate:
        players = getPlayers(cur, active=True)
        kwds = ("disable", "active")
        activate = False
    elif args.enableexpansion:
        expans = getExpans(cur, active=False)
        kwds = ("enable", "inactive")
        activate = True
    elif args.disableexpansion:
        expans = getExpans(cur, active=True)
        kwds = ("disable", "active")
        activate = False

    # handle expansion toggling
    if args.enableexpansion or args.disableexpansion:
        expnumlist = []
        for expan in expans:
            expnumlist.append(int(expan[0]))
            sys.stdout.write("{0:1.0f}) ".format(expan[0]) + expan[1] + "\n")
        try:
            toggleexp = input("Please enter the " + kwds[1] + ' expansion to ' + kwds[0] + ': ')
            if not list:
                sys.stderr.write("No expansions specified. Exiting.\n")
        except (EOFError, KeyboardInterrupt):
            conn.close()
            sys.exit()
        try:
            toggleexp = int(toggleexp)
        except:
            sys.stderr.write("Error: invalid input. Please enter a number from the list next time.\n")
            sys.exit()
        if toggleexp not in expnumlist:
            sys.stderr.write("Error: invalid input. Please enter a number from the list next time.\n")
            sys.exit()
        toggleExpan(cur, toggleexp, activate=activate)

    # handle player toggling
    if args.playeractivate or args.playerdeactivate:
        playernumlist = []
        for player in players:
            playernumlist.append(int(player[0]))
            sys.stdout.write("{0:1.0f}) ".format(player[0]) + player[1] + "\n")
        try:
            toggleplayer = input("Please enter the " + kwds[1] + ' player to ' + kwds[0] + ': ')
            if not list:
                sys.stderr.write("No player specified. Exiting.\n")
        except (EOFError, KeyboardInterrupt):
            conn.close()
            sys.exit()
        try:
            toggleplayer = int(toggleplayer)
        except:
            sys.stderr.write("Error: invalid input. Please enter a number from the list next time.\n")
            sys.exit()
        if toggleplayer not in playernumlist:
            sys.stderr.write("Error: invalid input. Please enter a number from the list next time.\n")
            sys.exit()
        togglePlayer(cur, toggleplayer, activate=activate)

    conn.commit()
    conn.close()


if __name__ == "__main__":
    main()
